backend/internal/observability/publisher/publisher.go

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright (c) 2025, WSO2 LLC. (https://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+// Package publisher provides the event bus mechanism for the observability system.
+package publisher
+
+import (
+	"sync"
+	"time"
+
+	"github.com/asgardeo/thunder/internal/observability/event"
+	"github.com/asgardeo/thunder/internal/observability/metrics"
+	"github.com/asgardeo/thunder/internal/observability/subscriber"
+	"github.com/asgardeo/thunder/internal/system/log"
+)
+
+// CategoryPublisher is an event bus interface with category-based routing.
+// Events are published synchronously to all interested subscribers.
+type CategoryPublisher interface {
+	// Publish publishes an event immediately to all interested subscribers.
+	Publish(event *event.Event)
+
+	// Subscribe adds a subscriber (subscriber decides what categories it wants).
+	Subscribe(sub subscriber.Subscriber)
+
+	// Unsubscribe removes a subscriber.
+	Unsubscribe(sub subscriber.Subscriber)
+
+	// GetActiveCategories returns categories that have at least one subscriber.
+	GetActiveCategories() []event.EventCategory
+
+	// Shutdown gracefully shuts down the publisher.
+	Shutdown()
+}
+
+// categoryEventPublisher implements category-based event bus with smart routing.
+// This is a simple synchronous event bus - no queuing, no async processing.
+// Events are published directly to subscribers who can decide whether to process them.
+type categoryEventPublisher struct {
+	subscribers           map[string]subscriber.Subscriber // subscriberID -> subscriber
+	subscribersByCategory map[event.EventCategory][]string // category -> []subscriberIDs
+	mu                    sync.RWMutex
+	isShutdown            bool
+}
+
+const publisherLogComponent = "CategoryEventBus"
```

Comment:
Does logComponent value needs to be changed to identify easily as this is coming from publisher?

Reply:
Need to rename this. We should use some generic naming for this